icommands.c

```diff
@@ -98,6 +104,191 @@ enum CommandResult mutt_parse_icommand(/* const */ char *line, struct Buffer *er
+      mutt_buffer_add_printf(buf, "macro %s %s \"%s\"\n", menu->name, key_binding, tmp.data);
+    }
+  }
+}
+
+/**
+ * dump_menu - Dumps all the binds or macros maps of a menu into a buffer
+ * @param buffer     Output buffer
+ * @param menu       Menu to dump
+ * @param bind       If true it's :bind, else :macro
+ * @retval bool      true if menu is empty, false if not
+ */
+static bool dump_menu(struct Buffer *buf, struct Mapping *menu, bool bind)
+{
+  bool empty = true;
+  struct Keymap *map = NULL, *next = NULL;
+
+  for (map = Keymaps[menu->value]; map; map = next)
+  {
+    next = map->next;
+
+    if (bind && map->op != OP_MACRO)
+    {
+      empty = false;
+      dump_map(buf, menu, map, true);
+    }
+    else if (!bind && map->op == OP_MACRO)
+    {
+      empty = false;
+      dump_map(buf, menu, map, false);
+    }
+  }
+
+  return empty;
+}
+
+/**
+ * dump_all_menus - Dumps all the binds or macros inside every menu
+ * @param buf      Output buffer
+ * @param bind     If true it's :bind ::icommand, else :macro
+ */
+static void dump_all_menus(struct Buffer *buf, bool bind)
+{
+  bool empty;
+  for (int i = 0; i < MENU_MAX; i++)
+  {
+    const char *menu_name = mutt_map_get_name(i, Menus);
+    struct Mapping menu = { menu_name, i };
+
+    empty = dump_menu(buf, &menu, bind);
+
+    /* Add a new line for readability between menus. */
+    if (!empty && i < MENU_MAX - 1)
+      mutt_buffer_addch(buf, '\n');
+  }
+}
+
+/**
+ * icmd_bind - Parse 'bind' and 'macro' commands - Implements ::icommand_t
+ */
+static enum CommandResult icmd_bind(struct Buffer *buf, struct Buffer *s,
+                                    unsigned long data, struct Buffer *err)
+{
+  FILE *fpout = NULL;
+  char tempfile[PATH_MAX];
+  bool dump_all = false, bind = (data == 0);
+
+  if (!MoreArgs(s))
+    dump_all = true;
+  else
+    mutt_extract_token(buf, s, 0);
+
+  if (MoreArgs(s))
+  {
+    /* More arguments potentially means the user is using the
+     * ::command_t :bind command thus we delegate the task. */
+    return MUTT_CMD_ERROR;
+  }
+
+  mutt_mktemp(tempfile, sizeof(tempfile));
+  fpout = mutt_file_fopen(tempfile, "w");
+  if (!fpout)
+  {
+    mutt_buffer_printf(err, _("Could not create temporary file %s"), tempfile);
+    return MUTT_CMD_ERROR;
+  }
+
+  struct Buffer *filebuf = mutt_buffer_alloc(4096);
+  if (dump_all || mutt_str_strcasecmp(buf->data, "all") == 0)
+  {
+    dump_all_menus(filebuf, bind);
+  }
+  else
+  {
+    const int menu_index = mutt_map_get_value(buf->data, Menus);
+    if (menu_index == -1)
+    {
+      mutt_file_fclose(&fpout);
+      mutt_buffer_free(&filebuf);
+      mutt_buffer_printf(err, _("%s: no such menu"), buf->data);
+      return MUTT_CMD_ERROR;
+    }
+
+    struct Mapping menu = { buf->data, menu_index };
+    dump_menu(filebuf, &menu, bind);
+  }
+
+  fputs(filebuf->data, fpout);
+  fflush(fpout);
+  mutt_file_fclose(&fpout);
+  mutt_buffer_free(&filebuf);
+
+  if (mutt_file_check_empty(tempfile) == 1)
```

Comment:
We can do this check more efficiently by using `buffer_is_empty()` on `filebuf`
(and moving the check before the file write).